source/util/ilist_node.h

```diff
@@ -120,15 +185,46 @@ inline void IntrusiveNodeBase<NodeType>::InsertAfter(NodeType* pos) {
 template <class NodeType>
 inline void IntrusiveNodeBase<NodeType>::RemoveFromList() {
   assert(!this->is_sentinel_ && "Sentinel nodes cannot be moved around.");
-  assert(this->next_node_ != nullptr && "Cannot remove a node from a list if it is not in a list.");
-  assert(this->previous_node_ != nullptr && "Cannot remove a node from a list if it is not in a list.");
+  assert(this->IsInAList() &&
+      "Cannot remove a node from a list if it is not in a list.");
 
   this->next_node_->previous_node_ = this->previous_node_;
   this->previous_node_->next_node_ = this->next_node_;
   this->next_node_ = nullptr;
   this->previous_node_ = nullptr;
 }
 
+template<class NodeType>
+void IntrusiveNodeBase<NodeType>::ReplaceWith(NodeType* target) {
+  if (is_sentinel_) {
+    assert(target->next_node_ == target);
+    assert(this->is_sentinel_);
+  } else {
+    assert(IsInAList() && "Sentinel nodes must always be part of a list.");
+    assert(!this->is_sentinel_ &&
+        "Cannot turn a sentinel node into one that is not.");
+  }
+
+  if (this->next_node_ != this) {
+    target->next_node_ = this->next_node_;
+    target->previous_node_ = this->previous_node_;
+
+    target->next_node_->previous_node_ = static_cast<NodeType*>(target);
+    target->previous_node_->next_node_ = static_cast<NodeType*>(target);
+
+    if (!this->is_sentinel_) {
+      this->next_node_ = nullptr;
+      this->previous_node_ = nullptr;
+    } else {
+      this->next_node_ = static_cast<NodeType*>(this);
+      this->previous_node_ = static_cast<NodeType*>(this);
+    }
+  } else {
```

Comment:
Please add comments.

Reply:
Fixed up a bit.  Let me know what you think.